### PR TITLE
C#: Limit extracted compilation and extraction messages

### PR DESCRIPTION
**Pull request by tamasvajk**
This PR reduces the number of extraction messages and compiler diagnostics that get stored in the database. I've added some env variables that can fine tune the limits, but I haven't introduced an extractor option for them.

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/CompilerDiagnostic.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            var messageCount = messageCounts.AddOrUpdate(key, 1, (_, c) => c + 1);
+            if (messageCount > limit)
+            {
+                if (messageCount == limit + 1)
+                {
+                    Context.Extractor.Logger.LogWarning($"Stopped logging {key} compiler diagnostics after reaching {limit}");
+                }
+
+                return;
+            }
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/CompilerDiagnostic.cs:37]: Should we enforce this limit for buildless extraction only?

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/CompilerDiagnostic.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
+using System.Collections.Concurrent;
+using System.IO;
+using Semmle.Util;
+
+namespace Semmle.Extraction.CSharp.Entities
+{
+    internal class CompilerDiagnostic : FreshEntity
+    {
+        private static readonly int limit = EnvironmentVariables.TryGetExtractorOption<int>("COMPILER_DIAGNOSTIC_LIMIT") ?? 1000;
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/CompilerDiagnostic.cs:9]: Should we introduce an extractor option for this? What should be the name for the option?

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/CompilerDiagnostic.cs (hunk 1 above)
Comment (hvitved) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/CompilerDiagnostic.cs:37]: I think it's fine to do for traced extraction as well.

File: csharp/extractor/Semmle.Util/EnvironmentVariables.cs [hunk 3]
```diff
@@ -7,6 +9,16 @@ public class EnvironmentVariables
         public static string? GetExtractorOption(string name) =>
             Environment.GetEnvironmentVariable($"CODEQL_EXTRACTOR_CSHARP_OPTION_{name.ToUpper()}");
 
+        public static T? TryGetExtractorOption<T>(string name) where T : struct, INumberBase<T>
```
Comment (hvitved) [inline csharp/extractor/Semmle.Util/EnvironmentVariables.cs:12]: `TryGetExtractorNumberOption`?

**Review by hvitved — APPROVED** [review]
LGTM; one suggestion.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/CompilerDiagnostic.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        protected override void Populate(TextWriter trapFile)
+        {
+            // The below doesn't limit the extractor messages to the exact limit, but it's good enough.
+            var key = diagnostic.Id;
+            var messageCount = messageCounts.AddOrUpdate(key, 1, (_, c) => c + 1);
+            if (messageCount > limit)
+            {
+                if (messageCount == limit + 1)
+                {
+                    Context.Extractor.Logger.LogWarning($"Stopped logging {key} compiler diagnostics after reaching {limit}");
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/CompilerDiagnostic.cs:33]: @yo-h brought up today that we could log the total number of compiler errors for telemetry purposes.

File: csharp/extractor/Semmle.Extraction/Entities/ExtractionMessage.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+        protected override void Populate(TextWriter trapFile)
+        {
+            // The below doesn't limit the extractor messages to the exact limit, but it's good enough.
+            Interlocked.Increment(ref messageCount);
+            if (messageCount > limit)
+            {
+                if (messageCount == limit + 1)
+                {
+                    Context.Extractor.Logger.LogWarning($"Stopped logging extractor messages after reaching {limit}");
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction/Entities/ExtractionMessage.cs:28]: Could we log the total number of extractor messages for telemetry purposes by storing the message count at the end of the extraction?

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/CompilerDiagnostic.cs (hunk 4 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/CompilerDiagnostic.cs]: This is done in https://github.com/github/codeql/pull/15957/commits/205d6a3bc5e55e9a2e4cfa9e522cae1219c22461.

File: csharp/extractor/Semmle.Extraction/Entities/ExtractionMessage.cs (hunk 5 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction/Entities/ExtractionMessage.cs:28]: Skipping this for the time being.

**Comment by tamasvajk** [comment]
@hvitved Can you have a look at the new commits in this PR.
The DCA experiment is failing with OOM, but I think that's on variant 0. Also the results show that there are many messages that are not logged any longer. It's not just buildless that's impacted.

**Review by hvitved — APPROVED** [review]

**Pull request merged**